nifpga/session.py

```diff
@@ -390,8 +410,272 @@ def read(self):
+        self._delta = self._calculate_delta(bitfile_register.word_length,
+                                            bitfile_register.integer_word_length)
+
+    @property
+    def overflow(self):
+        """ This property should only set be whenever this register has overflow
+        enabled"""
+        return self._overflow
+
+    @overflow.setter
+    def overflow(self, value):
+        if self._overflow_enabled:
+            self._overflow = value
+
+    def _calculate_delta(self, word_length, integer_word_length):
+        return 2**(integer_word_length - word_length)
+
+    def read(self):
+        """ Reads the entire FXP number as an array U32 bits then combining
+        the array bits into a single datatype, then convert it into python
+        Decimal. The largest expecting transfer would be a 64 bit fixed point
+        with the overflow status enabled (making an array of length 3).
+
+        Returns:
+            data (DataType.value): The data in the register in a python list.
+        """
+        data = super(_FxpRegister, self).read()
+        data = self._combine_array_of_u32_into_one_value(data)
+        return self._convert_from_read_value_to_decimal(data)
+
+    def _combine_array_of_u32_into_one_value(self, data):
+        combinedData = 0
+        for index in range(0, len(self)):
+            combinedData = (combinedData << 32 * index) + data[index]
+        if len(self) > 1:
+            overflow_bit = 1 if self._overflow_enabled else 0
+            total_num_bits = self._word_length + overflow_bit
+            combinedData = combinedData >> (32 * len(self) - total_num_bits)
+        return combinedData
+
+    def _convert_from_read_value_to_decimal(self, data):
+        """ This function converts read value from a RIO device and sets
+        the overflow bit if enabled and returns a Decimal representation of
+        the fixed point. """
+        if self._overflow_enabled:
+            self.overflow = self._get_overflow_value(data)
+            data = self._remove_overflow_bit(data)
+        if self._signed:
+            data = self._integer_twos_comp(data)
+        return Decimal(data * self._delta)
+
+    def _get_overflow_value(self, data):
+        """ Mask out all the data within the word length, leaving the overflow
+        bit. If the result after masking the the word portion of the fixed
+        point is none zero that indicates the data read has overflown."""
+        mask = 2**(self._word_length)
+        if data & mask > 0:
+            return True
+        return False
+
+    def _remove_overflow_bit(self, data):
+        """ This helper function masks out all bits not inside the word length,
+        ultimately returning a value of data missing the overflow bit. """
+        return data & (2**(self._word_length) - 1)
+
+    def _integer_twos_comp(self, data):
+        if (data & (1 << (self._word_length - 1))) != 0:
+            data = data - (1 << self._word_length)
+        return data
```

Comment:
I'm not sure I understand this function. The function name suggests that it will return the 2's complement of data, but the code seems to only return the 2's complement if the input's MSB is 1. Please explain.

Reply:
After more thought, I guess this function accepts a number (data) encoded using 2's complement and returns a Python integer. Maybe the function should have a name like _twos_comp_to_integer. If you're following some Python convention for naming type conversion functions, feel free to ignore my suggestion.